core/domain/classifier_services.py

```
@@ -127,6 +127,102 @@ def classify_string_classifier_rule(state, normalized_answer):
+            classifier_domain.TrainingJobExplorationMapping(
+                job_dicts_list[job_id_index]['exp_id'],
+                job_dicts_list[job_id_index]['exp_version'],
+                job_dicts_list[job_id_index]['state_name'],
+                job_id))
+        job_exploration_mapping.validate()
+        job_exploration_mappings.append(job_exploration_mapping)
+
+    _create_multi_job_exploration_mappings(job_exploration_mappings)
+
+
+def create_job_exploration_mappings(exploration, state_names,
+                                    state_names_mapping):
+    """Creates new TrainingJobExplorationMappingModel instances for all the
+    state names passed into the function. The mapping is created from the
+    new state to the ClassifierTrainingJob of the old state. If there's been
+    a change in the state name, we retrieve the old state name and create the
+    mapping accordingly.
+    This method is called only from exp_services._save_exploration() method and
+    is never called from classifier_services._create_exploration().
+
+    Args:
+        exploration: Exploration. The Exploration domain object.
+        state_names: list(str). List of state names.
+        state_names_mapping: dict. Dict mapping new state names to their
+            corresponding state names in previous version.
+    """
+    exp_id = exploration.id
+    current_exp_version = exploration.version
+    old_exp_version = current_exp_version - 1
+
+    state_names_to_retrieve = []
+    for current_state_name in state_names:
+        old_state_name = state_names_mapping[current_state_name]
+        state_names_to_retrieve.append(old_state_name)
+    classifier_training_jobs = get_classifier_training_jobs(
+        exp_id, old_exp_version, state_names_to_retrieve)
+
+    job_exploration_mappings = []
+    for index, classifier_training_job in enumerate(classifier_training_jobs):
+        if classifier_training_job is None:
+            continue
```

So what actually happens in this case? Is a new job created for the state? Should an error be logged?

(Not saying that this is wrong; just asking to ensure we've thought about it.)

---

@seanlip This is just a measure to avoid errors. This case will mostly never happen. This corresponds to the case where the previous version for a state did not have a job. If the previous version did really not have a job, that state will never end up in this method as a state with unchanged answer group. It will be sent to the job creation method.

---

"Avoiding errors" means something different, which is to take active steps to prevent the error case from happening. Here, what you're doing, if an error occurs, is simply sweeping the error under the carpet.

So, my question remains. What happens if there's an unexpected error (due to some coding bug perhaps)? Should we raise an exception? Should we log an error? Both?

(Though perhaps it's worth saving whatever can be saved before actually raising the exception.)

---

@seanlip In the eventuality that the bug happens, we could raise an exception, but that would halt the execution for the remaining stuff. I'm not sure what you mean by logging an error, how does that work?

---

```
import logging
logging.error(...)
```

And yep, I agree with you re the exception thing. Let's do logging.error(). In the worst case some state will just be missing a classifier mapping / classifier, right, and will fall back to the default rules?

---

@seanlip . Yep, worst case can happen if the mapping/job is deleted after being created for the previous version. I've added the error log command.